File: backend/app/schemas/willard_delivery.py
"""
Schemas de WillardDelivery — salida de plomo a Willard (W1).

El precio solo existe en el tipo `venta`. Ahi la linea acepta `unit_price` XOR
`total_price` (patron #95 D8: Johana a veces tiene el total y no el unitario).
"""
from datetime import datetime
from decimal import Decimal
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.utils.dates import BusinessDate

DeliveryType = Literal["venta", "abono_bateria", "abono_material"]


# ---------------------------------------------------------------- Lineas ---

class WillardDeliveryLineCreate(BaseModel):
    material_id: UUID
    quantity: Decimal = Field(..., gt=0)
    scale_weight_kg: Optional[Decimal] = Field(
        None, gt=0, description="Opcional al capturar, obligatorio al revisar (#95)"
    )


class WillardDeliveryLinePrice(BaseModel):
    """Precio de una linea al liquidar — solo tipo `venta`."""
    line_id: UUID
    unit_price: Optional[Decimal] = Field(None, gt=0)
    total_price: Optional[Decimal] = Field(None, gt=0)

    @model_validator(mode="after")
    def _xor_price(self):
        if (self.unit_price is None) == (self.total_price is None):
            raise ValueError(
                "Indique precio unitario O valor total de la linea, no ambos ni ninguno"
            )
        return self


class WillardDeliveryLineResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    material_id: UUID
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    material_unit: str = "kg"
    quantity: Decimal
    unit: Optional[str] = None
    scale_weight_kg: Optional[Decimal] = None
    kg_lead_equivalent: Optional[Decimal] = None
    unit_cost: Optional[Decimal] = None
    unit_price: Optional[Decimal] = None
    total_price: Optional[Decimal] = None


# ------------------------------------------------------------- Cabecera ---

class WillardDeliveryCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    delivery_type: DeliveryType
    warehouse_id: UUID
    third_party_id: UUID
    date: BusinessDate
    driver_id: Optional[UUID] = None
    vehicle_id: Optional[UUID] = None
    invoice_number: Optional[str] = Field(None, max_length=50)
    remission_number: Optional[str] = Field(None, max_length=50)
    notes: Optional[str] = Field(None, max_length=1000)
    lines: list[WillardDeliveryLineCreate] = Field(..., min_length=1)


class WillardDeliveryUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    warehouse_id: Optional[UUID] = None
    date: Optional[BusinessDate] = None
    driver_id: Optional[UUID] = None
    vehicle_id: Optional[UUID] = None
    invoice_number: Optional[str] = Field(None, max_length=50)
    remission_number: Optional[str] = Field(None, max_length=50)
    notes: Optional[str] = Field(None, max_length=1000)
    lines: Optional[list[WillardDeliveryLineCreate]] = Field(None, min_length=1)


class WillardDeliveryLiquidate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    line_prices: list[WillardDeliveryLinePrice] = Field(
        default_factory=list,
        description="Solo tipo `venta`: una entrada por linea",
    )
    customer_id: Optional[UUID] = Field(
        None, description="Solo tipo `venta`: cliente de la venta derivada (default Willard)"
    )


class WillardDeliveryAnnul(BaseModel):
    reason: str = Field(..., min_length=3, max_length=500)


class WillardDeliveryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    delivery_number: int
    delivery_type: DeliveryType
    warehouse_id: UUID
    warehouse_name: Optional[str] = None
    third_party_id: UUID
    third_party_name: Optional[str] = None
    date: datetime
    driver_id: Optional[UUID] = None
    driver_name: Optional[str] = None
    vehicle_id: Optional[UUID] = None
    vehicle_plate: Optional[str] = None
    invoice_number: Optional[str] = None
    remission_number: Optional[str] = None
    notes: Optional[str] = None
    status: str

    reviewed_at: Optional[datetime] = None
    reviewed_by_name: Optional[str] = None
    liquidated_at: Optional[datetime] = None
    liquidated_ts: Optional[datetime] = None
    liquidated_by_name: Optional[str] = None
    annulled_reason: Optional[str] = None
    annulled_at: Optional[datetime] = None
    annulled_by_name: Optional[str] = None
    created_by_name: Optional[str] = None

    sale_id: Optional[UUID] = None
    sale_number: Optional[int] = None

    maquila_amount: Decimal = Decimal("0")
    freight_amount: Decimal = Decimal("0")
    plant_credit_amount: Decimal = Decimal("0")

    total_kg_lead: Decimal = Decimal("0")

    lines: list[WillardDeliveryLineResponse] = Field(default_factory=list)


class WillardDeliveryListResponse(BaseModel):
    items: list[WillardDeliveryResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/willard-delivery.ts
// Salidas de plomo a Willard (W1).
//
// ⚠️ Todo lo que el backend declara como Decimal llega como STRING (FastAPI
// serializa asi). Es el bloqueante (b) de #93: el tipo decia `number`, `acc + x`
// concatenaba texto y el resultado fue "NaN kg". Aca los montos se declaran
// `string | number` y se leen SIEMPRE con `num()`.

export type WillardDeliveryType = "venta" | "abono_bateria" | "abono_material";
export type WillardDeliveryStatus = "draft" | "reviewed" | "liquidated" | "annulled";

/** Coercion en la frontera: el tipo se cumple porque se fuerza al entrar. */
export function num(value: string | number | null | undefined): number {
  if (value === null || value === undefined) return 0;
  const n = typeof value === "number" ? value : parseFloat(value);
  return Number.isFinite(n) ? n : 0;
}

export const DELIVERY_TYPE_LABELS: Record<WillardDeliveryType, string> = {
  venta: "Venta",
  abono_bateria: "Abono a batería",
  abono_material: "Abono a material",
};

export const DELIVERY_TYPE_COLORS: Record<WillardDeliveryType, string> = {
  venta: "bg-emerald-100 text-emerald-800",
  abono_bateria: "bg-violet-100 text-violet-800",
  abono_material: "bg-cyan-100 text-cyan-800",
};

export const DELIVERY_STATUS_LABELS: Record<WillardDeliveryStatus, string> = {
  draft: "Registrada",
  reviewed: "Revisada",
  liquidated: "Liquidada",
  annulled: "Anulada",
};

export const DELIVERY_STATUS_COLORS: Record<WillardDeliveryStatus, string> = {
  draft: "bg-amber-100 text-amber-800",
  reviewed: "bg-blue-100 text-blue-800",
  liquidated: "bg-green-100 text-green-800",
  annulled: "bg-slate-200 text-slate-600",
};

export interface WillardDeliveryLine {
  id: string;
  material_id: string;
  material_code: string | null;
  material_name: string | null;
  material_unit: string;
  quantity: string | number;
  unit: string | null;
  scale_weight_kg: string | number | null;
  kg_lead_equivalent: string | number | null;
  unit_cost: string | number | null;
  unit_price: string | number | null;
  total_price: string | number | null;
}

export interface WillardDelivery {
  id: string;
  delivery_number: number;
  delivery_type: WillardDeliveryType;
  warehouse_id: string;
  warehouse_name: string | null;
  third_party_id: string;
  third_party_name: string | null;
  date: string;
  driver_id: string | null;
  driver_name: string | null;
  vehicle_id: string | null;
  vehicle_plate: string | null;
  invoice_number: string | null;
  remission_number: string | null;
  notes: string | null;
  status: WillardDeliveryStatus;

  reviewed_at: string | null;
  reviewed_by_name: string | null;
  liquidated_at: string | null;
  /** Instante real del clic — este SI lleva hora (#93/#87) */
  liquidated_ts: string | null;
  liquidated_by_name: string | null;
  annulled_reason: string | null;
  annulled_at: string | null;
  annulled_by_name: string | null;
  created_by_name: string | null;

  sale_id: string | null;
  sale_number: number | null;

  maquila_amount: string | number;
  freight_amount: string | number;
  plant_credit_amount: string | number;
  total_kg_lead: string | number;

  lines: WillardDeliveryLine[];
}

export interface WillardDeliveryListResponse {
  items: WillardDelivery[];
  total: number;
  page: number;
  page_size: number;
}

export interface WillardDeliveryLineCreate {
  material_id: string;
  quantity: string;
  scale_weight_kg?: string | null;
}

export interface WillardDeliveryCreate {
  delivery_type: WillardDeliveryType;
  warehouse_id: string;
  third_party_id: string;
  date: string;
  driver_id?: string | null;
  vehicle_id?: string | null;
  invoice_number?: string | null;
  remission_number?: string | null;
  notes?: string | null;
  lines: WillardDeliveryLineCreate[];
}

export interface WillardDeliveryUpdate {
  warehouse_id?: string;
  date?: string;
  driver_id?: string | null;
  vehicle_id?: string | null;
  invoice_number?: string | null;
  remission_number?: string | null;
  notes?: string | null;
  lines?: WillardDeliveryLineCreate[];
}

export interface WillardDeliveryLinePrice {
  line_id: string;
  unit_price?: string | null;
  total_price?: string | null;
}

export interface WillardDeliveryLiquidate {
  line_prices: WillardDeliveryLinePrice[];
  customer_id?: string | null;
}
